<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Genotype Data &mdash; saegus 0.1.2 documentation</title>
    
    <link rel="stylesheet" href="../_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '0.1.2',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="top" title="saegus 0.1.2 documentation" href="../contents.html" />
   
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9">

  </head>
  <body role="document">  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="genotype-data">
<h1>Genotype Data<a class="headerlink" href="#genotype-data" title="Permalink to this headline">¶</a></h1>
<div class="section" id="handling-missing-genotype-data">
<h2>Handling missing genotype data<a class="headerlink" href="#handling-missing-genotype-data" title="Permalink to this headline">¶</a></h2>
<p>In most cases genotype-by-sequencing will have some missing data. This is a
tutorial for handling missing genotype data.</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="kn">import</span> <span class="nn">numpy</span> <span class="kn">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="kn">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">h5py</span> <span class="kn">as</span> <span class="nn">hp</span>
<span class="kn">import</span> <span class="nn">collections</span> <span class="kn">as</span> <span class="nn">col</span>
<span class="kn">import</span> <span class="nn">scipy.stats</span> <span class="kn">as</span> <span class="nn">st</span>
</pre></div>
</div>
<p>In this instance Randy wanted to include three more loci so I had to re-create a base population file to accommodate
the new genome structure. The new genotype matrix is in a file called &#8216;raw_genotype_matrix.txt&#8217; and the corresponding
genetic map in &#8216;genetic_map.txt&#8217;.</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">genotype_matrix_filename</span> <span class="o">=</span> <span class="s">&#39;raw_genotype_matrix.txt&#39;</span>
<span class="n">genetic_map_filename</span> <span class="o">=</span> <span class="s">&#39;genetic_map.txt&#39;</span>
</pre></div>
</div>
<div class="code python highlight-python"><div class="highlight"><pre><span class="k">def</span> <span class="nf">parse_genotype_matrix</span><span class="p">(</span><span class="n">genotype_matrix_filename</span><span class="p">:</span> <span class="nb">str</span><span class="p">):</span>
    <span class="n">genotype_matrix</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">genotype_matrix_filename</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s">&#39;</span><span class="se">\t</span><span class="s">&#39;</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">low_memory</span><span class="o">=</span><span class="bp">False</span><span class="p">)</span>
    <span class="n">genotype_matrix</span> <span class="o">=</span> <span class="n">genotype_matrix</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s">&#39;popdata&#39;</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">genotype_matrix</span>
</pre></div>
</div>
<div class="code python highlight-python"><div class="highlight"><pre><span class="k">def</span> <span class="nf">parse_recombination_rates</span><span class="p">(</span><span class="n">genetic_map_filename</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Returns a list of crossover probabilities from a genetic map measured in centimorgans.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">genetic_map</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">genfromtxt</span><span class="p">(</span><span class="n">genetic_map_filename</span><span class="p">,</span> <span class="n">delimiter</span><span class="o">=</span><span class="s">&#39;</span><span class="se">\t</span><span class="s">&#39;</span><span class="p">,</span> <span class="n">usecols</span><span class="o">=</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">3</span><span class="p">])</span>
    <span class="n">recombination_rates</span> <span class="o">=</span> <span class="n">col</span><span class="o">.</span><span class="n">OrderedDict</span><span class="p">()</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">genetic_map</span><span class="p">),</span> <span class="mi">1</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">genetic_map</span><span class="p">[</span><span class="n">i</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="n">genetic_map</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="mi">0</span><span class="p">]:</span>
            <span class="n">recombination_rates</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">divide</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">genetic_map</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span> <span class="o">-</span> <span class="n">genetic_map</span><span class="p">[</span><span class="n">i</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="mi">1</span><span class="p">]),</span> <span class="mi">100</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">genetic_map</span><span class="p">[</span><span class="n">i</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span> <span class="o">!=</span> <span class="n">genetic_map</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="mi">0</span><span class="p">]:</span>
            <span class="n">recombination_rates</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="mf">0.0</span>
    <span class="n">recombination_rates</span><span class="p">[</span><span class="nb">len</span><span class="p">(</span><span class="n">genetic_map</span><span class="p">)]</span> <span class="o">=</span> <span class="mf">0.0</span>
    <span class="k">return</span> <span class="nb">list</span><span class="p">(</span><span class="n">recombination_rates</span><span class="o">.</span><span class="n">values</span><span class="p">())</span>
</pre></div>
</div>
<div class="code python highlight-python"><div class="highlight"><pre><span class="k">def</span> <span class="nf">end_pt_finder</span><span class="p">(</span><span class="n">genetic_map_filename</span><span class="p">,</span> <span class="n">number_of_chromosomes</span><span class="o">=</span><span class="mi">10</span><span class="p">):</span>
    <span class="n">genetic_map</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">genetic_map_filename</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s">&#39;</span><span class="se">\t</span><span class="s">&#39;</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="bp">None</span><span class="p">)</span>
    <span class="n">genetic_map</span><span class="o">.</span><span class="n">drop</span><span class="p">([</span><span class="s">&#39;locus&#39;</span><span class="p">,</span> <span class="s">&#39;agpv2&#39;</span><span class="p">,</span> <span class="s">&#39;namZmPRDA&#39;</span><span class="p">,</span> <span class="s">&#39;namZmPRDS&#39;</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
    <span class="n">genetic_map</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">genetic_map</span><span class="p">)</span>
    <span class="n">end_pts</span> <span class="o">=</span> <span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">genetic_map</span><span class="p">)):</span>
        <span class="k">if</span> <span class="n">genetic_map</span><span class="p">[</span><span class="n">i</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span> <span class="o">!=</span> <span class="n">genetic_map</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="mi">0</span><span class="p">]:</span>
            <span class="n">end_pts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">pass</span>
    <span class="n">end_pts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">genetic_map</span><span class="p">))</span>
    <span class="n">chrom_lengths</span> <span class="o">=</span> <span class="p">[(</span><span class="n">end_pts</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">-</span> <span class="n">end_pts</span><span class="p">[</span><span class="n">i</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">number_of_chromosomes</span><span class="o">+</span><span class="mi">1</span><span class="p">)]</span>
    <span class="k">del</span> <span class="n">end_pts</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="k">return</span> <span class="n">end_pts</span><span class="p">,</span> <span class="n">chrom_lengths</span>
</pre></div>
</div>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">endpts</span><span class="p">,</span> <span class="n">lens</span> <span class="o">=</span> <span class="n">end_pt_finder</span><span class="p">(</span><span class="s">&#39;genetic_map.txt&#39;</span><span class="p">)</span>
</pre></div>
</div>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">gmatrix</span> <span class="o">=</span> <span class="n">parse_genotype_matrix</span><span class="p">(</span><span class="s">&#39;raw_genotype_matrix.txt&#39;</span><span class="p">)</span>
</pre></div>
</div>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">garray</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">gmatrix</span><span class="p">)</span>
</pre></div>
</div>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">missing_loci</span> <span class="o">=</span> <span class="p">[</span><span class="n">i</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">garray</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span> <span class="k">if</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span> <span class="ow">in</span> <span class="n">geno_cnt</span><span class="p">[</span><span class="n">i</span><span class="p">]]</span>
</pre></div>
</div>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">geno_cnt</span> <span class="o">=</span> <span class="p">{</span><span class="n">i</span><span class="p">:</span><span class="n">col</span><span class="o">.</span><span class="n">Counter</span><span class="p">(</span><span class="n">garray</span><span class="p">[:,</span> <span class="n">i</span><span class="p">])</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">garray</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">])}</span>
</pre></div>
</div>
<div class="code python highlight-python"><div class="highlight"><pre><span class="k">def</span> <span class="nf">genotype_counts_to_frequencies</span><span class="p">(</span><span class="n">genotype_counts</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Converts a the dictionaries of genotype: count for each locus into their</span>
<span class="sd">    frequency equivalents by dropping and missing data and dividing by the adjusted</span>
<span class="sd">    total.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">geno_frq</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">mlocus</span> <span class="ow">in</span> <span class="n">missing_loci</span><span class="p">:</span>
        <span class="n">geno_frq</span><span class="p">[</span><span class="n">mlocus</span><span class="p">]</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">if</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span> <span class="ow">in</span> <span class="n">geno_cnt</span><span class="p">[</span><span class="n">mlocus</span><span class="p">]:</span>
            <span class="k">del</span> <span class="n">geno_cnt</span><span class="p">[</span><span class="n">mlocus</span><span class="p">][</span><span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">]</span>
        <span class="n">inds_counted</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="n">geno_cnt</span><span class="p">[</span><span class="n">mlocus</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">())</span>
        <span class="k">for</span> <span class="n">genotype</span><span class="p">,</span> <span class="n">cnt</span> <span class="ow">in</span> <span class="n">geno_cnt</span><span class="p">[</span><span class="n">mlocus</span><span class="p">]</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="n">geno_frq</span><span class="p">[</span><span class="n">mlocus</span><span class="p">][</span><span class="n">genotype</span><span class="p">]</span> <span class="o">=</span> <span class="n">cnt</span><span class="o">/</span><span class="n">inds_counted</span>
    <span class="k">return</span> <span class="n">geno_frq</span>
</pre></div>
</div>
<div class="code python highlight-python"><div class="highlight"><pre><span class="k">def</span> <span class="nf">generate_genotype_pmfs</span><span class="p">(</span><span class="n">genotype_frequencies</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    For the time being all of the information required to compute a custom</span>
<span class="sd">    probability mass function for each locus is stored a dictionary keyed by locus.</span>
<span class="sd">    The values are tuples:</span>
<span class="sd">    0: genotype: frequency</span>
<span class="sd">    1: integer: genotype</span>
<span class="sd">    2: density</span>
<span class="sd">    3: genotype: integer</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">genotype_pmfs</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">loci</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">genotype_frequencies</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>
    <span class="k">for</span> <span class="n">missinglocus</span> <span class="ow">in</span> <span class="n">loci</span><span class="p">:</span>
        <span class="n">genotype_pmfs</span><span class="p">[</span><span class="n">missinglocus</span><span class="p">]</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">genotype_pmfs</span><span class="p">[</span><span class="n">missinglocus</span><span class="p">][</span><span class="s">&#39;states&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">geno_state</span> <span class="k">for</span> <span class="n">geno_state</span> <span class="ow">in</span> <span class="n">genotype_frequencies</span><span class="p">[</span><span class="n">missinglocus</span><span class="p">]</span><span class="o">.</span><span class="n">keys</span><span class="p">()]</span>
        <span class="n">genotype_pmfs</span><span class="p">[</span><span class="n">missinglocus</span><span class="p">][</span><span class="s">&#39;frequencies&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">frequency</span> <span class="k">for</span> <span class="n">frequency</span> <span class="ow">in</span> <span class="n">genotype_frequencies</span><span class="p">[</span><span class="n">missinglocus</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">()]</span>
        <span class="c"># scipy.stats.rv_discrete only allows integer or float valued random variables.</span>
        <span class="c"># So we have to map an integer to each genotype state value and vice versa.</span>
        <span class="n">genotype_pmfs</span><span class="p">[</span><span class="n">missinglocus</span><span class="p">][</span><span class="s">&#39;state_to_integer&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span><span class="n">state</span><span class="p">:</span><span class="n">i</span> <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">state</span>
                                                           <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">genotype_pmfs</span><span class="p">[</span><span class="n">missinglocus</span><span class="p">][</span><span class="s">&#39;states&#39;</span><span class="p">])}</span>
        <span class="n">genotype_pmfs</span><span class="p">[</span><span class="n">missinglocus</span><span class="p">][</span><span class="s">&#39;integer_to_state&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span><span class="n">i</span><span class="p">:</span><span class="n">state</span> <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">state</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">genotype_pmfs</span><span class="p">[</span><span class="n">missinglocus</span><span class="p">][</span><span class="s">&#39;states&#39;</span><span class="p">])}</span>
        <span class="n">integer_states</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">genotype_pmfs</span><span class="p">[</span><span class="n">missinglocus</span><span class="p">][</span><span class="s">&#39;state_to_integer&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">())</span>
        <span class="n">genotype_pmfs</span><span class="p">[</span><span class="n">missinglocus</span><span class="p">][</span><span class="s">&#39;pmf&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">st</span><span class="o">.</span><span class="n">rv_discrete</span><span class="p">(</span><span class="n">values</span><span class="o">=</span><span class="p">(</span><span class="n">integer_states</span><span class="p">,</span> <span class="n">genotype_pmfs</span><span class="p">[</span><span class="n">missinglocus</span><span class="p">][</span><span class="s">&#39;frequencies&#39;</span><span class="p">]))</span>
    <span class="k">return</span> <span class="n">genotype_pmfs</span>
</pre></div>
</div>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">gf</span> <span class="o">=</span> <span class="n">genotype_counts_to_frequencies</span><span class="p">(</span><span class="n">geno_cnt</span><span class="p">)</span>
</pre></div>
</div>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">gframe</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">garray</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="n">gmatrix</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">gmatrix</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span>
</pre></div>
</div>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">gframe</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s">&#39;genotype_matrix.txt&#39;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s">&#39;</span><span class="se">\t</span><span class="s">&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="../contents.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Genotype Data</a><ul>
<li><a class="reference internal" href="#handling-missing-genotype-data">Handling missing genotype data</a></li>
</ul>
</li>
</ul>
<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="../contents.html">Documentation overview</a><ul>
  </ul></li>
</ul>
</div>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="../_sources/tutorials/missing_genotype_data.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2016, John J. Dougherty III.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 1.3.1</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.6</a>
      
      |
      <a href="../_sources/tutorials/missing_genotype_data.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>