{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import simuOpt\n",
    "simuOpt.setOptions(alleleType='lineage', quiet=True)\n",
    "import simuPOP as sim\n",
    "import simuPOP.utils\n",
    "import simuPOP.lutils\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.patches as mpatches\n",
    "import math\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import random\n",
    "import datetime as dt\n",
    "import itertools as ite\n",
    "import seaborn as sns\n",
    "sns.set(context='paper', style='ticks', palette=\"Set2\", font='sans-serif')\n",
    "nucleotides = ['A','C','G','T','D','I']\n",
    "nucleotideTranslator = {\n",
    "    'A':0,\n",
    "    'C':1,\n",
    "    'G':2,\n",
    "    'T':3,\n",
    "    '-':4,\n",
    "    '+':5}\n",
    "from wgs import breed, operators, parameterizer, parser\n",
    "parse = parser.Parser()\n",
    "param = parameterizer.Parameterizer()\n",
    "f1Crosses = breed.fOne()\n",
    "f2Crosses = breed.fTwo()\n",
    "qtl = parameterizer.QTL()\n",
    "ae = parameterizer.AE()\n",
    "ld = parameterizer.LD()\n",
    "mavs = pd.DataFrame()\n",
    "def preparePopulation():\n",
    "    '''Simulation uses genetic maps from the NAM population.'''\n",
    "    N = 26 # There are 26 founders in the NAM population\n",
    "    # hapmap3.txt is not an actual 'HapMap'. Poor naming choice\n",
    "    header, mapdata, mapsubset=parse.mapParser('hapmap3.txt') \n",
    "    ends, chromdistances = parse.endPtFinder(mapsubset)\n",
    "    founder_population=sim.Population(N, loci=chromdistances, \n",
    "                                 alleleNames=nucleotides, \n",
    "                                 infoFields=['ind_id',\n",
    "                                            'father_id','mother_id',\n",
    "                                            'ge','pe','fitness'])\n",
    "    # neat.txt is a formatted output of hapmap3.txt. It is redundant\n",
    "    # and I need to clean this up some day.\n",
    "    param.genotypeSetter(founder_population,'neat.txt')\n",
    "    # defaults to founders 1 and 6\n",
    "    f1 = f1Crosses.fOneCross(founder_population, size=1, female_idx=1, male_idx=6)\n",
    "\n",
    "    M = 2000 # f2 population size\n",
    "    \n",
    "    f2 = f2Crosses.fTwoCross(f1, M)\n",
    "    # defaults to choosing qtl from segregating sites only\n",
    "    absQTL, propQTL = qtl.segChooser(f2, 20)\n",
    "    f2.dvars().qtl = absQTL\n",
    "    f2.dvars().properQTL = propQTL\n",
    "    alleleEffects, plotAlleleEffects = ae.exponential(f2, 1)\n",
    "    f2.dvars().alleleEffects = alleleEffects\n",
    "    f2.dvars().plottingAlleleEffects = plotAlleleEffects\n",
    "    sim.tagID(f2, reset=True)\n",
    "    # Parameters that are not used at this time. Will be integrated in\n",
    "    # later versions of WGS.\n",
    "    \n",
    "    #selection_intensity = .05\n",
    "    #f2.dvars().selectionIntensity = selection_intensity\n",
    "    #number_of_selected = f2.popSize() * selection_intensity\n",
    "    #remainder = f2.popSize() - number_of_selected\n",
    "    #number_of_breeding_subpops = int(number_of_selected / 5)\n",
    "    #list_of_applicable_subpops = list(range(1, number_of_breeding_subpops))\n",
    "    exec('import pandas as pd', f2.vars(), f2.vars())\n",
    "    exec('import itertools as ite', f2.vars(), f2.vars())\n",
    "    exec('import datetime as dt', f2.vars(), f2.vars())\n",
    "    exec('import matplotlib.pyplot as plt', f2.vars(), f2.vars())\n",
    "    # clone() has to be used here because it is a local variable\n",
    "    return f2.clone()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "f2 = preparePopulation()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def configureEvolution(pop, gens_to_evolve, initial_pop_size, proportion_selected, \n",
    "            overshoot_as_proportion,\n",
    "            individuals_per_breeding_subpop=5):\n",
    "    \"\"\"         *Notation n_* means 'number of*\n",
    "    pop = sim.Population\n",
    "    gens_to_evolve = evolves population for this number of generations\n",
    "    initial_pop_size = evolution usually begins with the output of an\n",
    "        f2 cross/random mating\n",
    "    proportion_selected = proportion of individuals to save at top of phenotypic\n",
    "        distribution (flexibility will be added in later versions).\n",
    "    overshoot_as_proportion = current breeding scheme involves randomly\n",
    "        selecting a subset of individuals from a larger pool of individuals\n",
    "    individuals_per_breeding_subpop = splits breeding_subpop into groups\n",
    "        of this many individuals\n",
    "    \n",
    "    Calculations of determined parameters from independent parameters.\n",
    "    I no longer have to specify \n",
    "    '\"\"\"\n",
    "    n_of_breeding_individuals = int(proportion_selected * initial_pop_size)\n",
    "    total_n_individuals_bred = int(initial_pop_size * (1 + overshoot_as_proportion))\n",
    "    n_of_breeding_subpops = int(initial_pop_size / n_of_breeding_individuals)\n",
    "    n_offspring_per_female =\\\n",
    "         int(total_n_individuals_bred / \\\n",
    "             (individuals_per_breeding_subpop*n_of_breeding_subpops))\n",
    "    n_bred_per_subpop = int(n_offspring_per_female*individuals_per_breeding_subpop)\n",
    "    n_of_nonbreeding_individuals = int(initial_pop_size - n_of_breeding_individuals)\n",
    "    n_of_individuals_removed = int(overshoot_as_proportion*initial_pop_size)\n",
    "    \n",
    "    pop.dvars().proportion_selected = proportion_selected\n",
    "    pop.dvars().individuals_per_breeding_subpop = individuals_per_breeding_subpop\n",
    "    pop.dvars().n_of_breeding_subpops = n_of_breeding_subpops\n",
    "    pop.dvars().n_of_nonbreeding_individuals = n_of_nonbreeding_individuals\n",
    "    pop.dvars().n_bred_per_subpop = n_bred_per_subpop\n",
    "    pop.dvars().n_of_individuals_removed = n_of_individuals_removed\n",
    "    \n",
    "    pc = breed.PC(individuals_per_breeding_subpop, n_offspring_per_female)\n",
    "    \n",
    "    pop.evolve(\n",
    "        initOps=[\n",
    "        operators.InitGenerationToZero(),\n",
    "        operators.InitializeMultiGenDataMatrix([\n",
    "                                      'smean_ge','svar_ge','smean_pe',\n",
    "                                      'svar_pe','umean_ge','uvar_ge',\n",
    "                                      'umean_pe','uvar_pe']),\n",
    "        operators.ParameterizationWriter(dt.datetime.now(),\n",
    "                                         'allelic-effects',\n",
    "                                         'for-plot-allelic-effects',\n",
    "                                         'improper-qtl',\n",
    "                                         'proper-qtl'),\n",
    "        sim.Stat(popSize=True),\n",
    "        sim.PyEval(r\"'Gen\\tMeanGe\\tVarGe\\tMeanPe\\tVarPe\\tSubPopulations\\n'\"),\n",
    "        ],\n",
    "        preOps=[\n",
    "            operators.GenotypicEffectCalculator(g_effect_field='ge'),\n",
    "            operators.TruncationSelection(0.7,proportion_selected, infoFields=['ge','pe']),\n",
    "            sim.Stat(meanOfInfo=['ge','pe'], varOfInfo=['ge','pe']),\n",
    "            sim.Stat(alleleFreq=sim.ALL_AVAIL),\n",
    "            #sim.Stat(numOfSegSites=sim.ALL_AVAIL),\n",
    "            #sim.Stat(effectiveSize=sim.ALL_AVAIL),\n",
    "            #sim.Stat(neutrality=sim.ALL_AVAIL),\n",
    "            sim.Stat(structure=sim.ALL_AVAIL, vars=['F_st']),\n",
    "            operators.GenerateMinorAlleleList(),\n",
    "            operators.MinorAlleleCounter(),\n",
    "            sim.PyEval(r\"'\\n%d\\t%.3f\\t%.3f\\t' % (gen, meanOfInfo['ge'], varOfInfo['ge'])\"),\n",
    "            sim.PyEval(r\"'%.3f\\t%.3f\\t%s\\n' % (meanOfInfo['pe'], varOfInfo['pe'], subPopSize)\"),\n",
    "            #operators.UniqueFileWriter(dt.datetime.now(), 'alleleFrequencies'),\n",
    "            operators.Sorter('pe'),\n",
    "            sim.SplitSubPops(sizes=[individuals_per_breeding_subpop]*n_of_breeding_subpops\\\n",
    "                            +[n_of_nonbreeding_individuals], randomize=False),\n",
    "            operators.CalcSplitMavs('ge','pe','smean_ge','svar_ge','smean_pe','svar_pe',\n",
    "                'umean_ge','uvar_ge','umean_pe','uvar_pe'),\n",
    "            ],\n",
    "        matingScheme=\n",
    "                sim.HomoMating(\n",
    "                    sim.PyParentsChooser(pc.recursivePairwiseParentChooser),\n",
    "                    sim.OffspringGenerator(ops=[\n",
    "                        sim.Recombinator(rates=0.01),\n",
    "                        sim.IdTagger(),\n",
    "                        sim.PedigreeTagger()],\n",
    "                        numOffspring=1),\n",
    "                    subPopSize=[n_bred_per_subpop]*n_of_breeding_subpops+[0],\n",
    "                    subPops=list(range(1,n_of_breeding_subpops)),\n",
    "                    ),\n",
    "        postOps=[\n",
    "            sim.MergeSubPops(),\n",
    "            operators.SelectRandomSeed(n_of_individuals_removed),\n",
    "            ],\n",
    "    gen=gens_to_evolve)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Gen\tMeanGe\tVarGe\tMeanPe\tVarPe\tSubPopulations\n",
      "\n",
      "0\t133.701\t47.173\t132.934\t466.782\t[2000]\n",
      "\n",
      "1\t138.500\t41.553\t138.424\t347.596\t[2000]\n",
      "\n",
      "2\t142.171\t29.846\t142.168\t195.011\t[2000]\n",
      "\n",
      "3\t146.127\t18.108\t146.190\t76.408\t[2000]\n",
      "\n",
      "4\t149.801\t12.402\t149.714\t39.428\t[2000]\n"
     ]
    }
   ],
   "source": [
    "configureEvolution(f2, 5, 2000, 0.05, \n",
    "            0.50,\n",
    "            individuals_per_breeding_subpop=5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>smean_ge</th>\n",
       "      <th>svar_ge</th>\n",
       "      <th>smean_pe</th>\n",
       "      <th>svar_pe</th>\n",
       "      <th>umean_ge</th>\n",
       "      <th>uvar_ge</th>\n",
       "      <th>umean_pe</th>\n",
       "      <th>uvar_pe</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>138.585875</td>\n",
       "      <td>44.497755</td>\n",
       "      <td>195.703410</td>\n",
       "      <td>31.003474</td>\n",
       "      <td>138.850342</td>\n",
       "      <td>114.582525</td>\n",
       "      <td>189.009337</td>\n",
       "      <td>0.432937</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>142.139838</td>\n",
       "      <td>19.995224</td>\n",
       "      <td>188.694422</td>\n",
       "      <td>6.262716</td>\n",
       "      <td>145.622255</td>\n",
       "      <td>1.942393</td>\n",
       "      <td>184.331668</td>\n",
       "      <td>0.555896</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>148.396361</td>\n",
       "      <td>6.474507</td>\n",
       "      <td>181.342447</td>\n",
       "      <td>1.359502</td>\n",
       "      <td>145.424744</td>\n",
       "      <td>10.855235</td>\n",
       "      <td>178.865578</td>\n",
       "      <td>0.542155</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>151.799147</td>\n",
       "      <td>7.379740</td>\n",
       "      <td>173.375101</td>\n",
       "      <td>3.826812</td>\n",
       "      <td>150.807200</td>\n",
       "      <td>17.754763</td>\n",
       "      <td>169.547915</td>\n",
       "      <td>0.922180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>154.393234</td>\n",
       "      <td>7.492218</td>\n",
       "      <td>168.398954</td>\n",
       "      <td>2.135985</td>\n",
       "      <td>152.969812</td>\n",
       "      <td>4.564765</td>\n",
       "      <td>165.795816</td>\n",
       "      <td>0.231533</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     smean_ge    svar_ge    smean_pe    svar_pe    umean_ge     uvar_ge  \\\n",
       "0  138.585875  44.497755  195.703410  31.003474  138.850342  114.582525   \n",
       "1  142.139838  19.995224  188.694422   6.262716  145.622255    1.942393   \n",
       "2  148.396361   6.474507  181.342447   1.359502  145.424744   10.855235   \n",
       "3  151.799147   7.379740  173.375101   3.826812  150.807200   17.754763   \n",
       "4  154.393234   7.492218  168.398954   2.135985  152.969812    4.564765   \n",
       "\n",
       "     umean_pe   uvar_pe  \n",
       "0  189.009337  0.432937  \n",
       "1  184.331668  0.555896  \n",
       "2  178.865578  0.542155  \n",
       "3  169.547915  0.922180  \n",
       "4  165.795816  0.231533  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f2.dvars().mavs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.4.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
