<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>analyze &mdash; saegus 0.1.2 documentation</title>
    
    <link rel="stylesheet" href="_static/nature.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '0.1.2',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="top" title="saegus 0.1.2 documentation" href="contents.html" />
    <link rel="prev" title="simulate" href="simulate.html" /> 
  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="simulate.html" title="simulate"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="contents.html">saegus 0.1.2 documentation</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="analyze">
<h1><code class="xref py py-mod docutils literal"><span class="pre">analyze</span></code><a class="headerlink" href="#analyze" title="Permalink to this headline">¶</a></h1>
<dl class="function">
<dt id="allele_data">
<code class="descname">allele_data</code><span class="sig-paren">(</span><em>pop</em>, <em>alleles</em>, <em>loci</em><span class="sig-paren">)</span><a class="headerlink" href="#allele_data" title="Permalink to this definition">¶</a></dt>
<dd><p>Determines the minor alleles, minor allele frequencies, major alleles and
major allele frequencies.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>pop</strong> &#8211; Population intended for GWAS analysis</li>
<li><strong>loci</strong> (<em>list</em>) &#8211; Loci for which to calculate frequency</li>
<li><strong>alleles</strong> (<em>dict</em>) &#8211; Dictionary of alleles present at each locus</li>
</ul>
</td>
</tr>
</tbody>
</table>
<p>This function is used to find the major/minor alleles of a Population
<code class="docutils literal"><span class="pre">pop</span></code> given a list of <code class="docutils literal"><span class="pre">alleles</span></code> at each locus given in <code class="docutils literal"><span class="pre">loci</span></code>.
The output is intended to be used in other functions to determine the
kinship matrix and population structure.</p>
<p>Additionally this function will also resolve ties between the
major and minor alleles which result when both alleles have exactly equal
frequency i.e. 0.50.</p>
</dd></dl>

<div class="highlight-python"><div class="highlight"><pre><span></span><span class="n">pop</span> <span class="o">=</span> <span class="n">sim</span><span class="o">.</span><span class="n">loadPopulation</span><span class="p">(</span><span class="s1">&#39;magic1478.pop&#39;</span><span class="p">)</span>
<span class="n">loci</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="n">pop</span><span class="o">.</span><span class="n">totNumLoci</span><span class="p">()))</span>
<span class="n">alleles</span> <span class="o">=</span> <span class="n">shelve</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="s1">&#39;magic_1478_simulation_parameters&#39;</span><span class="p">)</span>
<span class="n">alleles</span>

<span class="p">{</span><span class="mi">0</span><span class="p">:</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">],</span>
 <span class="mi">1</span><span class="p">:</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">3</span><span class="p">],</span>
 <span class="mi">2</span><span class="p">:</span> <span class="p">[</span><span class="mi">3</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span>
 <span class="mi">3</span><span class="p">:</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">2</span><span class="p">],</span>
 <span class="mi">4</span><span class="p">:</span> <span class="p">[</span><span class="mi">2</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span>
 <span class="mi">5</span><span class="p">:</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">2</span><span class="p">],</span>
 <span class="mi">6</span><span class="p">:</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">2</span><span class="p">],</span>
 <span class="mi">7</span><span class="p">:</span> <span class="p">[</span><span class="mi">3</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span>
 <span class="mi">8</span><span class="p">:</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">2</span><span class="p">],</span>
 <span class="o">...</span><span class="p">}</span>

 <span class="n">af</span> <span class="o">=</span> <span class="n">analyze</span><span class="o">.</span><span class="n">allele_data</span><span class="p">(</span><span class="n">magic1478_2718</span><span class="p">,</span> <span class="n">alleles</span><span class="p">,</span> <span class="nb">list</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="mi">1478</span><span class="p">)))</span>
 <span class="n">af</span>
</pre></div>
</div>
<div>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>minor_allele</th>
      <th>minor_frequency</th>
      <th>major_allele</th>
      <th>major_frequency</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>2</td>
      <td>0.00000</td>
      <td>1</td>
      <td>1.00000</td>
    </tr>
    <tr>
      <th>1</th>
      <td>3</td>
      <td>0.13275</td>
      <td>1</td>
      <td>0.86725</td>
    </tr>
    <tr>
      <th>2</th>
      <td>1</td>
      <td>0.06575</td>
      <td>3</td>
      <td>0.93425</td>
    </tr>
    <tr>
      <th>3</th>
      <td>2</td>
      <td>0.00000</td>
      <td>0</td>
      <td>1.00000</td>
    </tr>
    <tr>
      <th>4</th>
      <td>0</td>
      <td>0.05675</td>
      <td>2</td>
      <td>0.94325</td>
    </tr>
    <tr>
      <th>5</th>
      <td>2</td>
      <td>0.24875</td>
      <td>0</td>
      <td>0.75125</td>
    </tr>
    <tr>
      <th>6</th>
      <td>2</td>
      <td>0.12300</td>
      <td>0</td>
      <td>0.87700</td>
    </tr>
    <tr>
      <th>7</th>
      <td>1</td>
      <td>0.00000</td>
      <td>3</td>
      <td>1.00000</td>
    </tr>
    <tr>
      <th>8</th>
      <td>2</td>
      <td>0.24000</td>
      <td>0</td>
      <td>0.76000</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
  </tbody>
</table>
<p>1478 rows × 4 columns</p>
</div><dl class="function">
<dt id="rank_allele_effects">
<code class="descname">rank_allele_effects</code><span class="sig-paren">(</span><em>pop</em>, <em>loci</em>, <em>alleles</em>, <em>allele_effects</em><span class="sig-paren">)</span><a class="headerlink" href="#rank_allele_effects" title="Permalink to this definition">¶</a></dt>
<dd><p>Collects information about alleles at quantitative trait loci into a
dictionary. Determines favorable/unfavorable allele and corresponding
frequency. Keys of quantitative_trait_alleles have similar hierarchy
for both the alleles and their frequencies.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>pop</strong> &#8211; </li>
<li><strong>loci</strong> &#8211; </li>
<li><strong>alleles</strong> &#8211; </li>
<li><strong>allele_effects</strong> &#8211; </li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="allele_frq_table">
<code class="descname">allele_frq_table</code><span class="sig-paren">(</span><em>pop</em>, <em>number_gens</em>, <em>allele_frq_data</em>, <em>recombination_rates</em>, <em>genetic_map</em><span class="sig-paren">)</span><a class="headerlink" href="#allele_frq_table" title="Permalink to this definition">¶</a></dt>
<dd><p>Tabulates useful information about each locus and allele frequency</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>pop</strong> &#8211; Population with multiple sub-populations. Usually represents multiple generations of recurrent selection or drift.</li>
<li><strong>number_gens</strong> (<em>int</em>) &#8211; Number of generations of selection or drift</li>
<li><strong>allele_frq_data</strong> (<em>dict</em>) &#8211; Allele frequency data and the major/minor alleles at each locus.</li>
<li><strong>recombination_rates</strong> (<em>list</em>) &#8211; Recombination rates for each locus in order.</li>
<li><strong>genetic_map</strong> &#8211; Chromosome:cM position correspondence.</li>
</ul>
</td>
</tr>
</tbody>
</table>
<ul class="simple">
<li>abs_idex: Absolute index of locus</li>
<li>chrom: Chromosome</li>
<li>locus: Relative index of locus</li>
<li>major: Major allele</li>
<li>minor: Minor allele</li>
<li>recom_rate: Probability of recombination immediately <em>after</em> locus</li>
<li>cM: centiMorgan position on genetic map</li>
<li>v: miniature diagram where <cite>*</cite> is a breakpoint and <cite>|</cite> is a non-recombining locus</li>
<li>generation_labels: Generation number prefixed by <cite>G</cite>, frequency of minor allele at locus</li>
</ul>
<table border="1" class="docutils">
<colgroup>
<col width="14%" />
<col width="9%" />
<col width="9%" />
<col width="9%" />
<col width="9%" />
<col width="16%" />
<col width="8%" />
<col width="6%" />
<col width="6%" />
<col width="6%" />
<col width="6%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">abs_index</th>
<th class="head">chrom</th>
<th class="head">locus</th>
<th class="head">major</th>
<th class="head">minor</th>
<th class="head">recom_rate</th>
<th class="head">cM</th>
<th class="head">v</th>
<th class="head">G_0</th>
<th class="head">G_2</th>
<th class="head">G_4</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>0</td>
<td>1</td>
<td>0</td>
<td>0</td>
<td>2</td>
<td>0.00</td>
<td>-4.8</td>
<td><cite>|</cite></td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
</tr>
<tr class="row-odd"><td>1</td>
<td>1</td>
<td>1</td>
<td>2</td>
<td>0</td>
<td>0.01</td>
<td>-4.6</td>
<td><cite>|</cite></td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
</tr>
<tr class="row-even"><td>2</td>
<td>1</td>
<td>2</td>
<td>2</td>
<td>0</td>
<td>0.00</td>
<td>-4.4</td>
<td><cite>*</cite></td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
</tr>
<tr class="row-odd"><td>3</td>
<td>1</td>
<td>3</td>
<td>1</td>
<td>3</td>
<td>0.00</td>
<td>-4.2</td>
<td><cite>|</cite></td>
<td>0.1</td>
<td>0.2</td>
<td>0.3</td>
</tr>
<tr class="row-even"><td>4</td>
<td>1</td>
<td>4</td>
<td>1</td>
<td>2</td>
<td>0.00</td>
<td>-4.0</td>
<td><cite>|</cite></td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
</tr>
<tr class="row-odd"><td>5</td>
<td>1</td>
<td>5</td>
<td>5</td>
<td>4</td>
<td>0.00</td>
<td>-3.8</td>
<td><cite>|</cite></td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
</tr>
<tr class="row-even"><td>6</td>
<td>1</td>
<td>6</td>
<td>2</td>
<td>0</td>
<td>0.01</td>
<td>-3.6</td>
<td><cite>|</cite></td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
</tr>
<tr class="row-odd"><td>7</td>
<td>1</td>
<td>7</td>
<td>1</td>
<td>3</td>
<td>0.00</td>
<td>-3.4</td>
<td><cite>*</cite></td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
</tr>
</tbody>
</table>
<div class="code python highlight-python"><div class="highlight"><pre><span></span>allele_data = analyze.Frq(pop, triplet_qtloci[0], alleles_by_locus, qt_allele_effects[0])
allele_frequencies = allele_data.allele_frequencies(pop, range(pop.totNumLoci())
allele_frequency_table = selection_qtd.allele_frq_table(pop, 10, allele_frq_data, recombination_rates,
                                                      genetic_map)
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="generate_allele_effects_table">
<code class="descname">generate_allele_effects_table</code><span class="sig-paren">(</span><em>qtl</em>, <em>founder_alleles</em>, <em>allele_effects</em><span class="sig-paren">)</span><a class="headerlink" href="#generate_allele_effects_table" title="Permalink to this definition">¶</a></dt>
<dd><p>Creates a simple pd.DataFrame for allele effects. Hard-coded
for bi-allelic case.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>qtl</strong> (<em>list</em>) &#8211; List of loci declared as QTL</li>
<li><strong>alleles</strong> (<em>np.array</em>) &#8211; Array of alleles at each locus</li>
<li><strong>allele_effects</strong> (<em>dict</em>) &#8211; Mapping of effects for alleles at each QTLocus</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<p><strong>Example</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="n">alleles</span>
<span class="n">array</span><span class="p">([[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">],</span>
     <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">3</span><span class="p">],</span>
     <span class="p">[</span><span class="mi">3</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span>
     <span class="o">...</span><span class="p">,</span>
     <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span>
     <span class="p">[</span><span class="mi">3</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span>
     <span class="p">[</span><span class="mi">3</span><span class="p">,</span> <span class="mi">1</span><span class="p">]],</span> <span class="n">dtype</span><span class="o">=</span><span class="n">int64</span><span class="p">)</span>

<span class="n">qtl</span>
<span class="p">[</span><span class="mi">44</span><span class="p">,</span> <span class="mi">103</span><span class="p">,</span> <span class="mi">168</span><span class="p">,</span> <span class="mi">340</span><span class="p">,</span> <span class="mi">488</span><span class="p">,</span> <span class="mi">639</span><span class="p">,</span> <span class="mi">737</span><span class="p">,</span> <span class="mi">819</span><span class="p">,</span> <span class="mi">981</span><span class="p">,</span> <span class="mi">1065</span><span class="p">]</span>

<span class="n">allele_effects</span>
<span class="p">{</span><span class="mi">44</span><span class="p">:</span> <span class="p">{</span><span class="mi">0</span><span class="p">:</span> <span class="mf">5.629446187924926</span><span class="p">,</span> <span class="mi">2</span><span class="p">:</span> <span class="mf">1.8962727055819322</span><span class="p">},</span>
<span class="mi">103</span><span class="p">:</span> <span class="p">{</span><span class="mi">0</span><span class="p">:</span> <span class="mf">1.3097813991257303</span><span class="p">,</span> <span class="mi">2</span><span class="p">:</span> <span class="mf">6.14070564290979</span><span class="p">},</span>
<span class="mi">168</span><span class="p">:</span> <span class="p">{</span><span class="mi">2</span><span class="p">:</span> <span class="mf">6.718096248082958</span><span class="p">,</span> <span class="mi">3</span><span class="p">:</span> <span class="mf">4.697238579652859</span><span class="p">},</span>
<span class="mi">340</span><span class="p">:</span> <span class="p">{</span><span class="mi">1</span><span class="p">:</span> <span class="mf">1.521689147484636</span><span class="p">,</span> <span class="mi">2</span><span class="p">:</span> <span class="mf">2.2131077852927032</span><span class="p">},</span>
<span class="mi">488</span><span class="p">:</span> <span class="p">{</span><span class="mi">1</span><span class="p">:</span> <span class="mf">2.512286137462885</span><span class="p">,</span> <span class="mi">3</span><span class="p">:</span> <span class="mf">2.486777318327935</span><span class="p">},</span>
<span class="mi">639</span><span class="p">:</span> <span class="p">{</span><span class="mi">0</span><span class="p">:</span> <span class="mf">1.1268072986309254</span><span class="p">,</span> <span class="mi">3</span><span class="p">:</span> <span class="mf">1.3391282487711016</span><span class="p">},</span>
<span class="mi">737</span><span class="p">:</span> <span class="p">{</span><span class="mi">0</span><span class="p">:</span> <span class="mf">1.4879865577936147</span><span class="p">,</span> <span class="mi">1</span><span class="p">:</span> <span class="mf">1.607534785598338</span><span class="p">},</span>
<span class="mi">819</span><span class="p">:</span> <span class="p">{</span><span class="mi">1</span><span class="p">:</span> <span class="mf">2.2153417608326986</span><span class="p">,</span> <span class="mi">3</span><span class="p">:</span> <span class="mf">0.20077940947200731</span><span class="p">},</span>
<span class="mi">981</span><span class="p">:</span> <span class="p">{</span><span class="mi">0</span><span class="p">:</span> <span class="mf">3.9513501430851568</span><span class="p">,</span> <span class="mi">3</span><span class="p">:</span> <span class="mf">1.78843909724396</span><span class="p">},</span>
<span class="mi">1065</span><span class="p">:</span> <span class="p">{</span><span class="mi">0</span><span class="p">:</span> <span class="mf">0.998194377898828</span><span class="p">,</span> <span class="mi">2</span><span class="p">:</span> <span class="mf">1.5139052352904945</span><span class="p">}}</span>

 <span class="n">aeframe</span>
</pre></div>
</div>
<div>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>locus</th>
      <th>alpha_allele</th>
      <th>alpha_effect</th>
      <th>beta_allele</th>
      <th>beta_effect</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>44</td>
      <td>0</td>
      <td>5.629446</td>
      <td>2</td>
      <td>1.896273</td>
    </tr>
    <tr>
      <th>1</th>
      <td>103</td>
      <td>0</td>
      <td>1.309781</td>
      <td>2</td>
      <td>6.140706</td>
    </tr>
    <tr>
      <th>2</th>
      <td>168</td>
      <td>2</td>
      <td>6.718096</td>
      <td>3</td>
      <td>4.697239</td>
    </tr>
    <tr>
      <th>3</th>
      <td>340</td>
      <td>2</td>
      <td>2.213108</td>
      <td>1</td>
      <td>1.521689</td>
    </tr>
    <tr>
      <th>4</th>
      <td>488</td>
      <td>3</td>
      <td>2.486777</td>
      <td>1</td>
      <td>2.512286</td>
    </tr>
    <tr>
      <th>5</th>
      <td>639</td>
      <td>0</td>
      <td>1.126807</td>
      <td>3</td>
      <td>1.339128</td>
    </tr>
    <tr>
      <th>6</th>
      <td>737</td>
      <td>1</td>
      <td>1.607535</td>
      <td>0</td>
      <td>1.487987</td>
    </tr>
    <tr>
      <th>7</th>
      <td>819</td>
      <td>1</td>
      <td>2.215342</td>
      <td>3</td>
      <td>0.200779</td>
    </tr>
    <tr>
      <th>8</th>
      <td>981</td>
      <td>0</td>
      <td>3.951350</td>
      <td>3</td>
      <td>1.788439</td>
    </tr>
    <tr>
      <th>9</th>
      <td>1065</td>
      <td>2</td>
      <td>1.513905</td>
      <td>0</td>
      <td>0.998194</td>
    </tr>
  </tbody>
</table>
</div></div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h4>Previous topic</h4>
  <p class="topless"><a href="simulate.html"
                        title="previous chapter"><code class="docutils literal"><span class="pre">simulate</span></code></a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/analyze.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="simulate.html" title="simulate"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="contents.html">saegus 0.1.2 documentation</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2016, John J. Dougherty III.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3.1.
    </div>
  </body>
</html>