<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>&lt;no title&gt; &mdash; saegus 0.1.2 documentation</title>
    
    <link rel="stylesheet" href="_static/nature.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '0.1.2',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="top" title="saegus 0.1.2 documentation" href="contents.html" /> 
  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="contents.html">saegus 0.1.2 documentation</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <p>#! usr/bin/python
__author__ = &#8216;John J. Dougherty III&#8217;
# -<em>- coding: utf-8 -</em>-
import numpy as np
import pandas as pd
import csv
import collections as col</p>
<dl class="docutils">
<dt>def map_subsetter(map_data):</dt>
<dd>map_subset = [map_data[i] for i in range(len(map_data)) if int(float(map_data[i][3])) == float(map_data[i][3])]
return map_subset</dd>
<dt>def genotype_parser(genotype_matrix_filename, columns_to_drop=None,</dt>
<dd><blockquote class="first">
<div>droppable_individuals=None):</div></blockquote>
<p>&#8220;&#8221;&#8221;
Specific to raw NAM files.
Unsure if this is worth keeping here.
However, I am unsure of how to implement a flexible parser module without assuming that the file has a specific
structure already.
Obtains marker data from a file which tabular data file the 26 inbred lines
Assumes that genotype matrix will be in a format of Individual x Genotype Format</p>
<blockquote>
<div>Locus1  Locus2  Locus3  ...</div></blockquote>
<p>Ind1
Ind2
Ind3
...
If there are any droppable columns/individuals specify that in the args.
Low memory set to
&#8220;&#8221;&#8221;
genotype_matrix = pd.read_csv(genotype_matrix_filename, sep=&#8217;t&#8217;, index_col=0, low_memory=False)
if droppable_individuals is not None:</p>
<blockquote>
<div>genotype_matrix = genotype_matrix.drop(droppable_individuals, axis=0)</div></blockquote>
<dl class="docutils">
<dt>if columns_to_drop is not None:</dt>
<dd>genotype_matrix = genotype_matrix.drop(columns_to_drop, axis=1)</dd>
</dl>
<p class="last">return genotype_matrix</p>
</dd>
<dt>def end_pt_finder(genetic_map_filename, number_of_chromosomes=10):</dt>
<dd><p class="first">genetic_map = pd.read_csv(genetic_map_filename, sep=&#8217;t&#8217;, index_col=None)
end_pts = [0]
for i in range(1, len(genetic_map)):</p>
<blockquote>
<div><dl class="docutils">
<dt>if genetic_map[i-1][0] != genetic_map[i][0]:</dt>
<dd>end_pts.append(i)</dd>
<dt>else:</dt>
<dd>pass</dd>
</dl>
</div></blockquote>
<p class="last">end_pts.append(len(genetic_map))
chrom_lengths = [(end_pts[i] - end_pts[i-1]) for i in range(1, number_of_chromosomes+1)]
del end_pts[0]
return end_pts, chrom_lengths</p>
</dd>
</dl>
<p># Both qtl and ae file readers/writers need serious reconstruction.</p>
<dl class="docutils">
<dt>def ae_file_reader(pop, filename):</dt>
<dd><p class="first">ae_dict = col.OrderedDict()
with open(filename, newline=&#8217;n&#8217;) as ae_file:</p>
<blockquote>
<div><p>ae_reader = csv.reader(ae_file, delimiter=&#8217;,&#8217;, quotechar=&#8217;&#8221;&#8217;)
for row in ae_reader:</p>
<blockquote>
<div>splitrows = row[0].split(&#8221;,&#8221;)
stripped_sublocus = splitrows[0].lstrip(&#8216;(&#8216;)
stripped_nucleotide = splitrows[1].rstrip(&#8216;)&#8217;)
sl = int(stripped_sublocus)
nt = int(stripped_nucleotide)
effect = float(row[1])
ae_dict[sl, nt] = effect</div></blockquote>
</div></blockquote>
<p class="last">pop.dvars().alleleEffects = ae_dict</p>
</dd>
<dt>def parse_recombination_rates(genetic_map_filename):</dt>
<dd><p class="first">&#8220;&#8221;&#8221;
Returns a list of crossover probabilities from a genetic map measured in centimorgans.
&#8220;&#8221;&#8221;
genetic_map = pd.read_csv(genetic_map_filename, sep=&#8217;t&#8217;, index_col=None)
genetic_map.drop([&#8216;locus&#8217;, &#8216;agpv2&#8217;, &#8216;namZmPRDA&#8217;, &#8216;namZmPRDS&#8217;], axis=1, inplace=True)
genetic_map = np.array(genetic_map)
recombination_rates = col.OrderedDict()
for i in range(1, len(genetic_map), 1):</p>
<blockquote>
<div><dl class="docutils">
<dt>if genetic_map[i-1][0] == genetic_map[i][0]:</dt>
<dd>recombination_rates[i] = np.divide(np.abs(genetic_map[i][1] - genetic_map[i-1][1]), 100)</dd>
<dt>elif genetic_map[i-1][0] != genetic_map[i][0]:</dt>
<dd>recombination_rates[i] = 0.0</dd>
</dl>
</div></blockquote>
<p class="last">recombination_rates[len(genetic_map)] = 0.0
return list(recombination_rates.values())</p>
</dd>
<dt>def parse_genotype_matrix(genotype_matrix_filename: str, columns_to_drop=&#8217;popdata&#8217;, index_of_first_nonfounder=105):</dt>
<dd><p class="first">genotype_matrix = pd.read_csv(genotype_matrix_filename, sep=&#8217;t&#8217;, index_col=0, low_memory=False)
droppable_individuals = list(genotype_matrix.index[index_of_first_nonfounder:])
genotype_matrix = genotype_matrix.drop(droppable_individuals, axis=0)
if columns_to_drop is not None:</p>
<blockquote>
<div>genotype_matrix = genotype_matrix.drop(columns_to_drop, axis=1)</div></blockquote>
<p class="last">return genotype_matrix</p>
</dd>
</dl>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/parse.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="contents.html">saegus 0.1.2 documentation</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2016, John J. Dougherty III.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3.1.
    </div>
  </body>
</html>